File: Sources/ClipboardCore/OCRSelectionState.swift
import Foundation

public struct OCRTextBlock: Equatable, Identifiable, Sendable {
    public let id: String
    public var text: String

    public init(id: String, text: String) {
        self.id = id
        self.text = text
    }
}

public struct OCRSelectionState: Equatable, Sendable {
    public var blocks: [OCRTextBlock]
    public private(set) var selectedIDs: Set<String>

    public init(blocks: [OCRTextBlock], selectedIDs: Set<String> = []) {
        self.blocks = blocks
        self.selectedIDs = selectedIDs
    }

    public var canCopySelection: Bool {
        !selectedIDs.isEmpty
    }

    public var selectedText: String {
        blocks
            .filter { selectedIDs.contains($0.id) }
            .map(\.text)
            .joined(separator: "\n")
    }

    public mutating func toggle(_ id: String) {
        if selectedIDs.contains(id) {
            selectedIDs.remove(id)
        } else {
            selectedIDs.insert(id)
        }
    }
}

File: ClipboardSS.Windows/src/ClipboardSS.Core/Selection/OcrSelectionState.cs
namespace ClipboardSS.Core.Selection;

public sealed record OcrTextBlock(string Id, string Text);

public sealed class OcrSelectionState
{
    public OcrSelectionState(
        IEnumerable<OcrTextBlock> blocks,
        IEnumerable<string>? selectedIds = null)
    {
        Blocks = blocks.ToArray();
        SelectedIds = selectedIds?.ToHashSet(StringComparer.Ordinal)
            ?? new HashSet<string>(StringComparer.Ordinal);
    }

    public IReadOnlyList<OcrTextBlock> Blocks { get; }
    public HashSet<string> SelectedIds { get; }
    public bool CanCopySelection => SelectedIds.Count > 0;
    public string SelectedText => string.Join(
        "\n",
        Blocks.Where(block => SelectedIds.Contains(block.Id)).Select(block => block.Text));

    public void Toggle(string id)
    {
        if (!SelectedIds.Remove(id))
        {
            SelectedIds.Add(id);
        }
    }
}
